flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeySerializationUtils.java

```diff
@@ -138,4 +138,12 @@ private static void writeVariableIntBytes(
 			value >>>= 8;
 		} while (value != 0);
 	}
+
+	public static byte[] serializeKeyGroup(int keyGroup, int keyGroupPrefixBytes) {
+		byte[] startKeyGroupPrefixBytes = new byte[keyGroupPrefixBytes];
```

Comment:
Maybe we can rather pass the `startKeyGroupPrefixBytes` array directly instead of creating it in every invocation from `keyGroupPrefixBytes`. Like that, the caller can reuse the same array.

Reply:
Good point! There is one problem with this PR, After confirming with RocksDB the `deleteRange()` is still a experimental feature of RocksDB...even thought I did some experiments on this `deleteRange()` in our case, and I did find any downside, I'm not sure whether we should still use the `deleteRange()`, but the `deleteRange()` should be used for the recovery of the incremental checkpoint definitely when it's stable. What do you think?

Reply:
Some info about the experiment I did: 

- I set `ReadOptions::ignore_range_deletions = true` to speed up the read performance, because we won't read any records that belong to the key-group we have deleted. 
- I only call the `deleteRange()` twice, because we will at most call it twice in the recovery of the incremental checkpoint.

Reply:
This is a very good question. But I think for as long as we consider rescaling incremental checkpoints itself as experimental, we can try to use the `deleteRange` and change it in case we experience any problems. Would that be ok?

Reply:
@StefanRRichter I think that makes a lot of sense! I will rebase the PR and ping you again, it's a bit outdated now...

Reply:
Cool, no problem, I am also completing this review soon. One general thing that I was wondering about: did you ever see the sstable ingestion feature? It would be super nice for the rescaling of incremental checkpoints if we could simply ingest the sstables from multiple checkpoints into one database and the just clip the range boundaries. Unfortunately, from what I have seen this only works for external sstables written bei the sstable writer (see here: https://github.com/facebook/rocksdb/wiki/Creating-and-Ingesting-SST-files). I wonder if there is any way to modify the sstables if incremental checkpoints to make them usable for ingestion, but maybe it is just completely impossible. I also found this interesting discussion that outlines another potential approach: https://github.com/facebook/rocksdb/issues/499. Any thoughts?

Reply:
Yes, I did notice the "sstable ingestion feature", and also did some experiment on it. You are right that currently the ingestion feature only works for the sstables written by the sstable writer. I tried to use the sstable writer to generate external sstables in parallel and ingest the sstables into the target db, but unfortunately the performance of the sstable writer is quite poor in RocksJava...I left the experiment conclusion in [FLINK-8845](https://issues.apache.org/jira/browse/FLINK-8845)(that is the reason why I took a step back to use the `WriteBatch` to speed up the recovery for full checkpoint), I pasted the comments below:

**Unfortunately, even though according to RocksDB wiki, the best way to load data into RocksDB is "Generate SST files (using SstFileWriter) with non-overlapping ranges in parallel and bulk load the SST files.". But after implementing this and test with a simple bench mark, I found that the performance is not that good as expected, it's almost the same or worst that as using Rocks.put(). After a bit analysis I found that when building SST it consumed a lot of time to create DirectSlice and currently we can't reuse the DirectSlice in java api. Even though in C++ this could help to get a outperformance result, but in java I think we can't use this to improve the performance currently (maybe somedays RocksDB might improve this to enable us get a approximate performance in java as using C++) ...**

And regarding to https://github.com/facebook/rocksdb/issues/499, If I'm not misunderstand, I think we might also can't use the `repairDB()` because we have many column families, and the other opinions in that thread is quite similar with the approach that I've tried to build the sstables in parallel and it turned out that it didn't work properly with Java API.

Reply:
Ok, that was also how I understood the discussions and docs. In that ase, let's proceed with this approach and I will finalize the review now.